### ISSUE TITLE
WasmFS Documentation

### ISSUE THREAD

**Issue by tlively** [emscripten-core/emscripten#15949]
WasmFS should be well documented, with docs covering this non-exhaustive list of topics:

- [ ] How to build with WasmFS
- [ ] How to configure various backends and reason about when caches, etc. would be helpful.
- [ ] How to write your own backend.

**Comment by curiousdannii** [comment]
My project has two things that I'd like to use WasmFS for, so it would be great if the docs explained both of these:

1. A custom async FS. I have a custom FS currently, but it's hampered by not being async
2. Async stdin, which I currently implement by wrapping `getc` with an `EM_ASYNC_JS` function. I'm lucky that the project I'm porting only accesses stdin via `getc`, as it would be messy to have to wrap lots of other functions too.

**Comment by tlively** [comment]
re: mounting, I expect we will have both C and JS APIs for that. The story for Async APIs hasn't quite been figured out, but that's tracked in #15964.

**Comment by curiousdannii** [comment]
@tlively #15964 is about a new async API, right? I was expecting that it would possible to make the standard posix FS API async with Asyncify.

**Comment by tlively** [comment]
Oh gotcha. Yes, we might have to do something to get Asyncify working as expected as well. (Or maybe it will Just Work, not sure). I filed #16053 to track that separately.

**Comment by nachoab** [comment]
Is there some example of compiling including OPFS? Setting WASMFS=1 doesn't seem to include it, even though it [looks like it is](https://github.com/emscripten-core/emscripten/blob/main/src/modules.js#L103). Also, using current FS you have to use the mount method to specify a different FS provider, does WasmFS use this same API?

**Comment by mere-human** [comment]
> Is there some example of compiling including OPFS?

@nachoab I think the best example would be tests, for example https://github.com/emscripten-core/emscripten/blob/main/test/wasmfs/wasmfs_opfs.c

**Comment by curiousdannii** [comment]
It's taken a long while, but I'm ready to start working on my own WasmFS backend.

How "ready" is WasmFS now? I know some people are using it, but I also see lots of open issues and tickets in the project/milestone. It seems like there's still not really any docs, we just have to learn from the other implementations?

I suspect it might not be that difficult to use WasmFS with Asyncify, because you can put *anything* through Asyncify. The question is more about what's efficient and ergonomic.

An alternative idea I had was to somehow replace stdio.h with a custom implementation, but that would probably be much more complex.

**Comment by tlively** [comment]
We have some very large applications using WasmFS, especially with the OPFS backend, so it's "ready" for production use in the sense that it is stable and correct enough. We haven't been able to implement as many new features as we wanted, though, such as other concrete backends (for IndexedDB especially) and virtual backends for caching and layering, etc. We also haven't been able to finish all the compatibility work that we want to have done to allow us to turn on WasmFS by default.

**Comment by art926** [comment]
Could you update 
https://emscripten.org/docs/api_reference/Filesystem-API.html#file-systems
to mention OPFS, please?) We're very interested in using it.

**Comment by adamziel** [comment]
> We have some very large applications using WasmFS, especially with the OPFS backend, so it's "ready" for production use in the sense that it is stable and correct enough.

@tlively I would love to explore using the OPFS backend in [WordPress Playground](https://github.com/WordPress/wordpress-playground/), but I'm not sure how to do it. Would that work without pthreads support? We're using MEMFS in web browsers and MEMFS+NODEFS in Node.js. Would you be able to point me to any resources, or write up a guide in a comment in this issue?

**Comment by jozefchutka** [comment]
I also came here to explore options for OPFS backend. Preferably would like to mount FileSystemDirectoryHandle as mentioned in https://github.com/emscripten-core/emscripten/issues/15964#issuecomment-2265602899 but have no idea how/where to start. Will OPFS backend be provided by emscripten or something a dev would implement himself following some documentation?

**Comment by adamziel** [comment]
Looping in @thomasthedane @tomayac – an OPFS backend would be a highly useful addition.

**Comment by kripken** [comment]
> I would love to explore using the OPFS backend in [WordPress Playground](https://github.com/WordPress/wordpress-playground/), but I'm not sure how to do it. Would that work without pthreads support?

I believe the OPFS backend requires threads or JS Promise Integration (JSPI). One of the two is needed so that we can handle the async behavior of the JS APIs there. Here are the tests that we run:

https://github.com/emscripten-core/emscripten/blob/a756ff0bd47e3acc364c2d221056e67ea8cfebf4/test/test_browser.py#L5205-L5225

The code in those tests could be an easy way to get started with the OPFS backend.

Separately, it would be good to update the docs on the website.

**Comment by adamziel** [comment]
Thanks @kripken! I started exploring and got some initial successes, I'll report back when I have something workable. It will probably take a few months of taking a few moments here and there to get this right like [it took 1,5 years to get JSPI support in](https://github.com/WordPress/wordpress-playground/issues/134).

It would be much easier to use this API if there was a high quality documentation page. Not scattered blog posts, issues, and commits, but an actual learning resource with detailed, step by step guides examples, descriptions of different wasmfs backends, specific terminal commands to run, outcomes to expect, mental models explained, and live demos of the built binaries. Also, a transition guide from the JavaScript FS to WASMFS would help a lot.

Right now we, the developers trying to use WebAssembly, are in a tough spot. It's very unclear what's in progress, what's finished, what's the browser support, and how to use it.

**Comment by kripken** [comment]
@adamziel I agree, this could be much improved. There is also the remaining work to get WasmFS enabled by default, and I was thinking that might happen before, but perhaps it doesn't need to be in that order. I'm not sure when I'll have time myself to get to either, but I'd like to (and others might get to it before).

fyi @tomayac , this could be a promising area for docs or a blogpost or both.

**Comment by tomayac** [comment]
I agree that there's a need to document this properly. Meanwhile, this [repository](https://github.com/milky2028/emscripten-fs-minimal) might help people bootstrap their apps. May I also recommend people working on this install my [OPFS Explorer](https://chromewebstore.google.com/detail/opfs-explorer/acndjpgkpaclldomagafnognkcgjignd) Chrome DevTools extension? The files that you open with this demo are persisted across reloads, as you can verify with the extension. 

![Screenshot 2024-10-10 at 11 02 26](https://github.com/user-attachments/assets/97609c41-3745-430a-a594-07090db2df82)

**Comment by adamziel** [comment]
Thanks @tomayac! I've forked that demo repository and added a few more examples:

https://github.com/adamziel/emscripten-wasmfs-demo

**Comment by nexryai** [comment]
I'm having trouble getting WASMFS (OPFS) to work with JSPI.
I've tried many build options, including `JSPI_EXPORTS`, but OPFS initialization fails with the following error:
```c++
wasmfs_create_directory("/opfs", 0755, opfs);
```
```
SuspendError: trying to suspend without WebAssembly.promising
```

The offending WASM binary code, as shown in the stack trace, is as follows:
```
call $env._wasmfs_opfs_init_root_directory
```

After searching Google and GitHub for a solution, I found that @adamziel also appears to be encountering the same issue.
https://github.com/adamziel/emscripten-wasmfs-demo/tree/b66cb7e0c8b8413d9e4a95483f623f134e504b32?tab=readme-ov-file#running-the-demo

Is this an Emscripten issue, or am I missing something?
Due to technical constraints, I can't use ptread, so JSPI is my only option.

My linkopts and copts are as follows.  
I actually set similar options in copts, but I've omitted them because they're redundant.
```python
cc_binary(
    name = "ikaria",
    linkopts = [
        "-O3",
        "--bind",
        "-sINITIAL_MEMORY=399966208",
        "-sEXPORT_ES6=1",
        "-sWASMFS",
        "-sASYNCIFY=2",
        "-sENVIRONMENT=web,worker",
        "-sJSPI_EXPORTS=['_wasmfs_opfs_init_root_directory']",
        "-sASYNCIFY_IGNORE_INDIRECT=1",
        "-sUSE_PTHREADS=0",
        "-sASSERTIONS",
        "-o ikaria.js",
    ],
    deps = [":ikaria_lib"],
)
```

This appears to be covered by the tests.

https://github.com/emscripten-core/emscripten/blob/1f46cb4bfa7f2c2fb4932345dde39a6b2e65f24c/test/wasmfs/wasmfs_opfs.c#L26-L33

https://github.com/emscripten-core/emscripten/blob/1f46cb4bfa7f2c2fb4932345dde39a6b2e65f24c/test/test_browser.py#L5306-L5321

I'm completely stuck, and any information would be a great help.

**Comment by nexryai** [comment]
After some trial and error, I finally found the cause.
It seems that an error occurs when using OPFS in a function bound with `EMSCRIPTEN_BINDINGS` rather than `main()`.
When I changed `wasmfs_create_directory` to run in the main function, the error disappeared.
If I still try to write to an OPFS-mounted directory in a function bound with EMSCRIPTEN_BINDINGS, I get the following error:

```
SuspendError: trying to suspend without WebAssembly.promising
at opfs.wasm:0x453c0c
```

```
0x453c0c call $env._wasmfs_opfs_get_child
```

Looking at the JavaScript that Emscripten is generating, I think this is because Emscripten doesn't recognize the function I bound as asynchronous.


```javascript
        instrumentWasmExports(e) {
            var t = /^(main|__main_argc_argv)$/;
            pr.asyncExports = new Set;
            var r = {};
            for (let[a,i] of Object.entries(e))
                if ("function" == typeof i) {
                    t.test(a) && (pr.asyncExports.add(i),
                    i = pr.makeAsyncFunction(i));
                    var n = pr.instrumentFunction(i);
                    r[a] = n
                } else
                    r[a] = i;
            return r
        },
        asyncExports: null,
        isAsyncExport: e => pr.asyncExports?.has(e),
        handleAsync: async e => {
            try {
                return await e()
            } finally {}
        }
        ,
        handleSleep: e => pr.handleAsync( () => new Promise(e)),
        makeAsyncFunction: e => WebAssembly.promising(e)
    };
```

https://github.com/emscripten-core/emscripten/blob/035c704d5cbeb85481eb1c3273617d21b881d674/src/lib/libasync.js#L156-L167

I've set the following bindings based on the documentation, but for some reason they're not working.

```c++
using namespace emscripten;

EMSCRIPTEN_BINDINGS(dash_remuxer_module) {
    class_<DashRemuxer>("DashRemuxer")
        .constructor<>()
        .function("process", &DashRemuxer::process, async());
}
```
https://emscripten.org/docs/porting/asyncify.html#usage-with-embind
